### ISSUE TITLE
Mentioning do not work when username ends with `.` and other special characters

### ISSUE THREAD

**Issue by gmsecrieru** [RocketChat/Rocket.Chat#956]
![screen shot 2015-10-02 at 08 02 13](https://cloud.githubusercontent.com/assets/190883/10245228/f0203d86-68db-11e5-9553-7b9199434e3b.png)
##

**Comment by AdrianoCahete** [comment]
Or change username policies to not let usernames starts or finish with dot.

**Comment by gmsecrieru** [comment]
_Updated_: Looks like the problem is not related only to dots:

![image](https://cloud.githubusercontent.com/assets/190883/10264349/7a6f3466-69df-11e5-9b72-b8777b83f0d1.png)

**Comment by geekgonecrazy** [comment]
@rocketchat/owners I guess the question we need to answer first is, do we want usernames ending in non alphanumeric characters?

I personally dislike usernames ending in a period or dash.  I find it confusing.

Ex: Hey have you seen @ninja. around?

That's just my two cents.

**Comment by sampaiodiego** [comment]
I agree with you guys.. we should not accept usernames ending with non alphanumeric characters.

**Comment by emcguinness** [comment]
From a new user who is looking into this for his company perspective I agree with geekgonecrazy, ending in special characters could cause confusion (common punctuation more than others).

````
Hi there @user, how are you?
````
Am I speaking to `@user` and using proper punctuation or am I skipping punctuation and talking to `@user'`

**Comment by ninja-** [comment]
problem is the algorithm which suggests new usernames if they are taken and that adds a dot

**Comment by geekgonecrazy** [comment]
@ninja- correct.  That's why I ask.  Cause we either fix this and continue to allow this or make tweaks to username creation.

**Comment by steedos** [comment]
What about UTF-8 usernames, especially for Chinese , Japanese and Korean, 

发自我的 iPhone

**Comment by engelgabriel** [comment]
Thats a good point @steedos, we have to create a global UTF-8 REGEX for usernames and channels that takes that in consideration.

@ninja- are you having problems with this?

**Comment by ninja-** [comment]
@engelgabriel , changed my username to remove a dot. Algorithm which replaced my github username from ninja- to ninja. has a room for improvement

**Comment by kyrofa** [comment]
Note that this issue also applies if the username does not end in a period. 

![screenshot from 2016-09-08 09-48-30](https://cloud.githubusercontent.com/assets/946674/18358342/74043d8e-75a9-11e6-99d8-e61d438ff531.png)

**Comment by geekgonecrazy** [comment]
@kyrofa yeah it looks like we need to get our regex around username matching fixed asap..

**Comment by thef1sh** [comment]
Adding a space as a valid character in the regex breaks the mentioning completely when using the autocomplete

**Comment by geekgonecrazy** [comment]
if node.js supported negative lookbacks with regex this would be as easy as appending `(?<![\.])` to the regexp.

Works in chrome, but does not work in node 8 or probably other browsers.

**Comment by ChameleonScales** [comment]
@thef1sh if adding a space as a valid character breaks the autocomplete, then space should simply be a forbidden character when creating your nickname.

**Comment by sketch242** [comment]
@ChameleonScales With Custom OAuth and Office 365, your only options for user names are "Full Name" or "user@email.domain", both of which break mentions.  Not sure if I should open another ticket for having an @ in the username breaking mentions as well, or just mention it here.

**Comment by ChameleonScales** [comment]
but just why even allow a character that breaks a feature ?
When you put a slash / in a filename inside a folder, you get a popup telling you that you can't do that, and you have to change the name. To me that's how it should behave.

**Comment by xzag** [comment]
@ChameleonScales @sketch242 Current "Names Validation" regexp is also used for channels and groups. Therefore you either have mentions working and no spaces, dots, commas, etc.. in usernames, groups and channels. Or you ignore mentions and everything else is working.

I'm not sure if mentions for complex usernames will ever be implemented but at least maybe it is possible to separate those regexes.

**Comment by ChameleonScales** [comment]
@sketch242 I'm not talking about removing a character or mangling any data. My request is that when you click OK when creating your username, before actually validating, some script would first check if there's a space in your name and if there is, it would not validate and show a message telling you why.
do you think this would be doable ?

**Comment by rafaelks** [comment]
Just to mention that this is affecting the mobile apps on sending messages as well: 

![39147218-71432946-470f-11e8-8a72-931ed2c77043](https://user-images.githubusercontent.com/551004/39184227-09951f0e-4799-11e8-8544-5d7359d69948.jpg)

**Comment by hugocostadev** [comment]
working now

**Comment by ulope** [comment]
Nope still broken when adding regular sentence punctuation after a mention:

![](https://ds.ulo.pe/Screen-Shot-2023-07-03-15-12-13.07-65fbZY.jpg)

**Comment by hugocostadev** [comment]
Thanks @ulope for updating this
